### ISSUE TITLE
Cholesky: return info instead of RuntimeError

### ISSUE THREAD

**Issue by twoertwein** [pytorch/pytorch#47608]
## 🚀 Feature
`cholesky` raises a `RuntimeError` if the matrix is not (numerically) positive-definite. This makes it impossible to use it with TorchScript (for loop over increasing regularization of try-except with cholesky).

Similar to `lu`, return `info` to indicate which of the batches matrices were successfully decomposed and which weren't.

## Motivation
Use `cholesky` with TorchScript.

## Pitch

Return `lu`-like `info` instead of raising a `RuntimeError`.

cc @vishwakftw @jianyuh @nikitaved @pearu @mruberry @heitorschueroff

**Comment by xwang233** [comment]
This is a nice feature and we are still discussing about returning a `info` instead of checking for valid output. See discussion https://github.com/pytorch/pytorch/issues/42666#issuecomment-716414847

Related #42666

**Comment by twoertwein** [comment]
@xwang233 how would one check whether the output is valid? The simplest way would be to set L[batch, 0, 0] to a negative number to indicate that Choleksy failed for this matrix (diagonal entries should be strictly positive).

I assume that there are more methods that raise errors. An `info` tensor might be more consistent compared to method-specific encodings of errors (especially if people do not check for the encoded errors).

**Comment by xwang233** [comment]
Yes, you are right. For checking valid output, I mean checking the `info` tensor. 

I would suggest an option that linear algebra functions return a `info` tensor and let user choose to check it or not (or maybe not return `info` anyway), instead of check `info` all the time. Is that also what you suggested?

**Comment by twoertwein** [comment]
If it doesn't decrease performance to create `info` (I assume it is just a 1-dim bool tensor with one entry for each batch entry), then it would be cool to always return it. I can't imagine a user not wanting to check whether the results are valid or just some random numbers.

I understand that this would break the API - an option to return `info` only when requested would make more sense from that perspective.

I'm more than happy with either solution :)

**Comment by xwang233** [comment]
Well, I should say this in the other way: if we don't check the `info` tensor, the performance may improve. Though we focus on different points about `info`, the final implementation could be the same (with `info` tensor returned explicitly and checked by user optionally). 😂

The API may be changed if `info` is returned or not checked. Since it's still under discussion, feel free to leave your comments there #42666

**Comment by mruberry** [comment]
Just fyi the plan is to continue to do the check in linalg.cholesky but we'll offer linalg.cholesky_ex that returns the cholesky result and the info tensor and does no error checking of its own.

cc @IvanYashchuk 

We will accept PRs implementing the _ex variants of linear algebra operations that perform their own error checking.

**Comment by ngimel** [comment]
#34272 is also related.

**Comment by twoertwein** [comment]
`torch.nn.cholesky_ex` works perfectly. Thank you!
